### ISSUE TITLE
Furthering code and project

### ISSUE THREAD

**Issue by wltjr** [thewaiter/Clipboard#15]
I am looking to further the code and project, and have [forked the repo](https://github.com/Obsidian-StudiosInc/clipboard). Not sure if you want me to send PRs. I dropped autotools for meson. Various code cleanups and fixes with more to come. I will be looking to remove all ecore_x specific code, replacing with elm_cnp. Making it no longer bound to X, so should work under Wayland as well as X. Also updating to new E modules API. Hopefully fixing a few issues along the way.

I have merged in changes from @rbtylee [fork](https://github.com/rbtylee/e21-modules-extra/tree/master/Clipboard). I did not see any change that made it specific to e21+ but maybe. My fork very well may work in e17/moskha as well as e21+ at this time, not sure. Though eventually will only be e21-22+. I am not sure if you want to maintain 1 repo with a couple branches. Or if you want me to just continue on with my forked version. Either way just wanted to make it known. Thanks for your work!

**Comment by thewaiter** [comment]
Hello hello :)

We are happy you contribute this project. We really need some e dev help we wished long time. Please continue your work on you fork until we check the code funcionality under Moksha. Our top priority is solving the main problem: I chose not very elegant way for checking the clipboard content. It works the way, the ecore timer asks every second if sth new is there. This is certainly not good. Could you help us with this issue at first, please?

**Comment by wltjr** [comment]
Great! Though I am not an e developer. I do not really relate well with some e devs... :) I have been working on a few apps;
- [ecrire](https://github.com/Obsidian-StudiosInc/ecrire) text editor
- [entrance](https://github.com/Obsidian-StudiosInc/entrance) login manager
- [eminence](https://github.com/Obsidian-StudiosInc/eminence) theme

At some point I will look to do some stuff to further [enjoy](https://git.enlightenment.org/apps/enjoy.git/) music player. I had discussed Clipboard with @rbtylee a few times. Just took me a while to get to the point of touching the code. Switched to meson since it speeds up build which saves time in development. Somewhat prefer cmake + ninja, but since not making say a deb or rpm, went with meson. I guess could make packages of this module just the same.

Thus far I have just made minor code fixes short of swapping out build systems. I will have to double check the changes @rbtylee made. I am not sure if any of that made it not work under Moksha. I think it sill will work under Moksha. May just need to be tested. Though long term I know that will change. Not sure if all the changes I make can can back ported for Moksha or not.

I will look into the issue you mentioned. I am still familiarizing myself with the code. I will see about starting with the ecore timer issue. Before I proceed with making changes that will definitely not work under Moksha. Likely be a bit before I start making changes that will cause it to not work under Moksha. Unless it already does not work under Moshka due to @rbtylee changes. But that could likely be reverted for Moshka keeping other improvements.

**Comment by thewaiter** [comment]
Great and thanx. 
Maybe you will succeed and we will be able porting the fixes to the Moksha.

**Comment by thewaiter** [comment]
BTW: The clipboard project is very small. Does it make sense to switch to another build system? Compiling takes just a few seconds. I understand this for EFL and E. It takes ages...

**Comment by wltjr** [comment]
Its mostly the configure portion that takes time. If you look with meson it takes like seconds. My biggest long term gripe with autotools is the time to run configure. I think actual compile via make vs meson difference is not that much but slightly. I still like cmake as it can generate autotools or ninja build files. Just a bit more familiar with meson atm. When your doing lots of builds, the time savings can add up even for small code.

The initial changes will be fine for Moshka, again providing it still works now less @rbtylee changes. But when I start moving to elm_cnp, that will need a newer EFL I believe. That should not break anything for Moshka. Not till I look to switch to the newer module API. Pretty sure gadget stuff is gone and has changed, just remains for legacy for now in E 22, but I think for e 23+ it maybe removed not sure.

At that point is when a decision needs to be made, as to either different branches. Or maybe a shared codebase with 2 UIs for old and new E API. Something a build system could switch between and keep it all in one. But I do not have any interest in Moshka. Others will have to test and stuff there. But I am not opposed to a codebase that works for both.

**Comment by wltjr** [comment]
By the way, I was not meaning to dismiss Moshka, just declaring my interest lies else where. Also regarding size of clipboard. To my surprise, it has almost as many [lines of code(62,268)](https://scan.coverity.com/projects/obsidian-studiosinc-clipboard) (link will be public soon) as [Entrance(62,418)](https://scan.coverity.com/projects/obsidian-studiosinc-entrance) and more than [Ecrire(54,369)](https://scan.coverity.com/projects/obsidian-studiosinc-ecrire). Granted Entrance had more I removed, and will have more added back and other stuff removed...

**Comment by wltjr** [comment]
@thewaiter I removed most all of the timer code. I am still working on moving to elm_cnp and removing any ecore_x specific code. The removing timer just happened as part of that. Seems the timer was running to poll X for any changes to clipboard. Basically grabbing clipboard contents and storing it in clipboard modules history. I am attempting to switch to mouse/keyboard events for such. Those may more rapidly at times, but over the long run. Should be less as it will only fire with interaction. Presently that is not working correctly. Pretty sure some code in clipboard_x is still helping out there, ecore_x code vs elm_cnp. Still a ways to go, more code to write, lots of testing, and more ecore_x code to remove. It is still working for me, but I am not confident my code is being called to add stuff to the history. I think that is still coming from ecore_x code in clipboard_x.

It was not the goal at the time, but seems in the process of moving to elm_cnp I may have addressed the timer issue you mentioned. Not sure if all features will carry over or not, syncing, etc. Long as clipboard history is added and retained between applications is my main need. X and wayland are a bit differently so no sure if the clipboard handling will be similar enough for further stuff in elm_cnp. Or if I can replicate all of it now.

I also switched from e_widget_menu to e_widget_button. I may go back to the menu, not sure. I kind of like the popup vs menu. The buttons a bit impartial but had problems with ilist, and not sure about callbacks. Button seemed to work out ok, but take up more space. They are centered vs left aligned and some other issues. That is not as important but I did that to address a shelf not hiding issue. Though I think that was more with how the menu was being positioned etc. I may restore the menu vs buttons.

**Comment by thewaiter** [comment]
Hello. If you mean the reply to my mail from 30.3.2018 I got 3.4.2018 then yes.

**Comment by wltjr** [comment]
@thewaiter yes, just confirming you received the reply and did not end up in your junk/spam folder :)
